src/server/svr_jobfunc.c

```diff
@@ -1470,6 +1471,176 @@ svr_chkque(job *pjob, pbs_queue *pque, char *hostname, int mtype)
 }
 
+/**
+ * @brief
+ *		check_block_wt	-	A worktask to reply to the blocked job client
+ *
+ * @param[in]	ptask	-	work_task structure
+ */
+void
+check_block_wt(struct work_task *ptask)
+{
+	struct block_job_reply *blockj = ptask->wt_parm1;
+	int time_difference;
+	time_t current_timestamp = time(NULL);
+	int is_connected = 0;
+	int is_replied = 0;
+	int is_error = 0;
+
+	if (blockj->ji_block_fd == -1) {
+		int sock_flags;
+		int	is_connect;
+		struct hostent		*hp;
+		struct sockaddr_in	remote;
+		short			remote_sin_family;
+
+		if ((hp = gethostbyname(blockj->ji_submit_host)) == NULL) {
+			sprintf(log_buffer, "%s: client host %s not found for block job %s", __func__,
+			blockj->ji_submit_host, blockj->ji_jobid);
+			log_event(PBSEVENT_ERROR, PBS_EVENTCLASS_JOB, LOG_NOTICE,
+				blockj->ji_jobid, log_buffer);
+			free(blockj->ji_msg);
+			free(blockj);
+			return;
+		}
+
+		remote_sin_family = hp->h_addrtype;
+		memset(&remote, 0, sizeof(remote));
+		memcpy(&remote.sin_addr, hp->h_addr, hp->h_length);
+		remote.sin_port = htons((unsigned short)blockj->ji_port);
+		remote.sin_family = remote_sin_family;
+
+		blockj->ji_block_fd = socket(AF_INET, SOCK_STREAM, 0);
+		/* Set socket to Non-blocking */
+		if (blockj->ji_block_fd != -1) {
+			sock_flags = fcntl(blockj->ji_block_fd, F_GETFL, 0);
+			if (fcntl(blockj->ji_block_fd, F_SETFL, sock_flags | O_NONBLOCK) == -1) {
+				close(blockj->ji_block_fd);
+				blockj->ji_block_fd = -1;
+			}
+		}
+
+		/* The flag "is_connected" indicates the readiness of the socket and could be one of the following:
+		1. If it is 0, the socket is not created at all.
+		2. If it is 1, the socket is connected successfully and message can be send now.
+		3. If it is 2, the socket connection is in progress and needs to be polled now.
+		*/
+		if (blockj->ji_block_fd != -1) {
+			is_connect = connect(blockj->ji_block_fd, (struct sockaddr *)&remote, sizeof(remote));
+			if (is_connect == -1) {
+				if (errno == EINPROGRESS || errno == ECONNREFUSED || errno == EADDRINUSE || (errno == ETIMEDOUT) ||
+					errno == ECONNREFUSED || errno == EWOULDBLOCK) {
+					is_connected = 2;
+				} else {
+					close(blockj->ji_block_fd);
+					blockj->ji_block_fd = -1;
+				}
+			} else {
+				is_connected = 1;
+			}
+		}
+	} else {
+		/* If the socket exists, then it's meant for poll */
+		is_connected = 2;
+	}
+
+	if (is_connected == 2) {
+		int write_fds;
+		struct pollfd fds[1];
+		fds[0].fd = blockj->ji_block_fd;
+		fds[0].events  = POLLOUT;
+		fds[0].revents = 0;
+		write_fds = poll(fds, (nfds_t)1, 0);
+		if (write_fds == 1) {
+			int rc = 0;
+			int check_error;
+			pbs_socklen_t len = sizeof(rc);
+			check_error = getsockopt(fds[0].fd, SOL_SOCKET, SO_ERROR, &rc, &len);
+			if ((rc == 0) && check_error == 0 && (fds[0].revents == POLLOUT)) {
+				is_connected = 1;
+			}
+		}
+		if (write_fds == -1) {
+			close(blockj->ji_block_fd);
+			blockj->ji_block_fd = -1;
+		}
+	}
+	
+	if (is_connected == 1) {
```

Comment:
Can we use "else if" here?

Reply:
As per our offline discussion, I keep it as it is since I am setting this flag in the upper if block as well.